standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java

```diff
@@ -20,71 +20,646 @@
+      dbt = DbType.ORACLE;
+    } else if (productName.contains(POSTGRESQL_NAME)) {
+      dbt = DbType.POSTGRES;
     } else {
-      return OTHER;
+      dbt = DbType.OTHER;
     }
+
+    // This method may be invoked by concurrent connections
+    synchronized (DatabaseProduct.class) {
+      if (theDatabaseProduct == null) {
+        theDatabaseProduct = new DatabaseProduct();
+      }
+  
+      theDatabaseProduct.dbType = dbt;
+    }
+    return theDatabaseProduct;
+  }
+
+  public final boolean isDERBY() {
+    return dbType == DbType.DERBY;
+  }
+
+  public final boolean isMYSQL() {
+    return dbType == DbType.MYSQL;
   }
 
-  public static boolean isDeadlock(DatabaseProduct dbProduct, SQLException e) {
+  public final boolean isORACLE() {
+    return dbType == DbType.ORACLE;
+  }
+
+  public final boolean isSQLSERVER() {
+    return dbType == DbType.SQLSERVER;
+  }
+
+  public final boolean isPOSTGRES() {
+    return dbType == DbType.POSTGRES;
+  }
+
+  public final boolean isEXTERNAL() {
+    return dbType == DbType.EXTERNAL;
+  }
+
+  public final boolean isOTHER() {
+    return dbType == DbType.OTHER;
+  }
+
+  public boolean isDeadlock(SQLException e) {
     return e instanceof SQLTransactionRollbackException
-        || ((dbProduct == MYSQL || dbProduct == POSTGRES || dbProduct == SQLSERVER)
+        || ((isMYSQL() || isPOSTGRES() || isSQLSERVER() || isEXTERNAL())
             && "40001".equals(e.getSQLState()))
-        || (dbProduct == POSTGRES && "40P01".equals(e.getSQLState()))
-        || (dbProduct == ORACLE && (e.getMessage() != null && (e.getMessage().contains("deadlock detected")
+        || (isPOSTGRES() && "40P01".equals(e.getSQLState()))
+        || (isORACLE() && (e.getMessage() != null && (e.getMessage().contains("deadlock detected")
             || e.getMessage().contains("can't serialize access for this transaction"))));
   }
 
   /**
    * Whether the RDBMS has restrictions on IN list size (explicit, or poor perf-based).
    */
-  public static boolean needsInBatching(DatabaseProduct dbType) {
-    return dbType == ORACLE || dbType == SQLSERVER;
+  protected boolean needsInBatching() {
+    return isORACLE() || isSQLSERVER();
   }
 
   /**
    * Whether the RDBMS has a bug in join and filter operation order described in DERBY-6358.
    */
-  public static boolean hasJoinOperationOrderBug(DatabaseProduct dbType) {
-    return dbType == DERBY || dbType == ORACLE || dbType == POSTGRES;
+  protected boolean hasJoinOperationOrderBug() {
+    return isDERBY() || isORACLE() || isPOSTGRES();
   }
 
-  public static String getHiveSchemaPostfix(DatabaseProduct dbType) {
+  public String getHiveSchemaPostfix() {
     switch (dbType) {
     case SQLSERVER:
       return "mssql";
     case DERBY:
     case MYSQL:
     case POSTGRES:
     case ORACLE:
+    case EXTERNAL:
       return dbType.name().toLowerCase();
     case OTHER:
     default:
       return null;
     }
   }
+
+  public static void reset() {
```

Comment:
Curious to know why we would need this? If we don't do this may be we can make theDatabaseProduct as a final variable.

Reply:
:-) ... It's so that we don't fall into the assertion you suggested right above (i.e. "Preconditions.checkState(this.dbType == getDbType(productName));").
There is a test in the Hive test suite that creates multiple instances of DatabaseProduct for different databases, which fails unless the instance is reset.
I had added the same assertion you suggested when I originally wrote the code, but then removed it when I got bitten by the test. I was on the fence of whether the assertion was useful or not, so I removed it. I now put it back, along with the calls to reset() in the test.

Reply:
Thanks for the context. In such a case I think it would be good to annotate this method as @VisibleForTesting